FAB_LED.h

```diff
@@ -953,57 +959,156 @@ avrBitbangLedStrip<FAB_TVAR>::eightPortSoftwareSendBytes(const uint16_t count, c
 	uint8_t r6 __asm__("r8") = 0;
 	uint8_t r7 __asm__("r9") = 0;
 
-	for(uint16_t c __asm__("r11") = 0; c < blockSize; c++) {
-		for(int8_t b __asm__("r12") = 7; b >= 0; b--) {
-			uint8_t bitmask __asm__("r10") = 0;
+	// Since bits for each port are sent delayed by one cycle to
+	// implement the memory access pipeline, we must leave the
+	// ports not yet in use set to LOW level. Once the pipeline
+	// is initialized fully onMask == 0xFF, and we always set it
+	// to high to send a zero or a one.
+	uint8_t onMask __asm__("r10") = 0;
 
-			bitmask |=  r0 & 1      ; r0 >>= 1;
-			bitmask |= (r1 & 1) << 1; r1 >>= 1;
-			bitmask |= (r2 & 1) << 2; r2 >>= 1;
-			bitmask |= (r3 & 1) << 3; r3 >>= 1;
-			bitmask |= (r4 & 1) << 4; r4 >>= 1;
-			bitmask |= (r5 & 1) << 5; r5 >>= 1;
-			bitmask |= (r6 & 1) << 6; r6 >>= 1;
-			bitmask |= (r7 & 1) << 7; r7 >>= 1;
+	for(uint16_t c __asm__("r12") = 0; c < blockSize + 1; ++c) {
+		for(int8_t b __asm__("r13") = 7; b >= 0; --b) {
+			uint8_t bitmask __asm__("r10") = 0;
 
 			// Load ONE byte per iteration
 			// Skip end condition check to reduce CPU cycles.
 			// It is expected that the LED strip won't have extra pixels.
-			if ((b >= dataPortPin) && (b <= clockPortPin)) switch (b) {
+			switch (b) {
 				case 0:
-					r0 = array[c];
+					// By checking statically if rX is read from memory, we
+					// avoid changing it, and the compiler will optimize out all
+					// the code including bitmask setting. On 16MHz Uno, we can
+					// only set bitmask for 6 ports before the strip resets.
+					if (0 == dataPortPin) {
+						onMask |= 1;
+						r0 = array[c];
+					}
+
+					if (r0 & 0b10000000) bitmask |= 1;
+					if (r1 & 0b01000000) bitmask |= 2;
+					if (r2 & 0b00100000) bitmask |= 4;
+					if (r3 & 0b00010000) bitmask |= 8;
+					if (r4 & 0b00001000) bitmask |= 16;
+					if (r5 & 0b00000100) bitmask |= 32;
+					if (r6 & 0b00000010) bitmask |= 64;
+					if (r7 & 0b00000001) bitmask |= 128;
 					break;
 				case 1:
-					r1 = array[c + 1 * blockSize];
+					if ((1 >= dataPortPin) && (1 <= clockPortPin)) {
```

Comment:
You mean for the port check?
This should not matter, as it is all static values. I expect the compiler to optimize all this away as either true (aka the memory access stays) or false (the if block vanishes, so no memory access, as well as all the bitcode attached to that port).

I'll look at inline asm (idk how to do it yet), to see if I can shorten the code generated for
if (x & mask) bitmask |= constant;

If I can save a cycle on that operation it will all fit for 8 ports. However I bet it already translates to something compact like:

 AND x, mask
 BZ end
 OR bitmask, constant
end:

So I'd need to get rid of the AND somehow...